### ISSUE TITLE
FR: more Gesture Manager actions

### ISSUE THREAD

**Issue by pickx** [koreader/koreader#4333]
The Gesture Manager is a great feature for making KOReader faster to use.

That said, here are features I use frequently and would love as options when a gesture is triggered:

- [x] Toggle landscape <-> portrait
- [ ] Open last document
- [x] Sleep device
- [x] History (of last opened files)

thanks!

**Comment by Biep** [comment]
In reference to the issue Frenzie just referenced:

Currently I am really wanting a "Return to where I jumped from" gesture (as opposed to a menu item) that is different from the "go to the previous page" one.  Often I jump, then move several times to the next page, and then want to jump back - or sometimes want to leaf back.  A single gesture cannot serve both functions.

More generally, I'd love (as a user) having the ability to assign gestures to specific commands - the way hotkeys can be assigned by the user in many programs.

**Comment by Frenzie** [comment]
The gesture manager is the early stage of exactly that btw. It's still developing, but before @robert00s added it there was nothing. It's slow acquiring more configurable gestures (e.g., short diagonal swipe) and actions.

**Comment by Frenzie** [comment]
A few quick notes while my attention is focused on this issue. (I was a bit busy in November.)

tl;dr I'll implement at least some of these after dinner.

## Toggle landscape <-> portrait
There's already a rotation gesture for that on devices that support two-finger gestures. I haven't looked into where it lives but it should be simple to reuse or otherwise jig up with `Device:setRotation()`.

## Open last document
Probably not too hard but currently implemented locally in the menu.

https://github.com/koreader/koreader/blob/6b3158e497228614591dfd654d6318ffffa27156/frontend/apps/reader/modules/readermenu.lua#L194-L205

https://github.com/koreader/koreader/blob/6b3158e497228614591dfd654d6318ffffa27156/frontend/apps/reader/modules/readermenu.lua#L219

## Sleep device
Huh, no physical buttons? Anyway, that's just `UIManager:suspend()`, couldn't be simpler.

## History (of last opened files)
Sending a `ShowHist` event might do the trick. I'll have to check.

**Comment by Frenzie** [comment]
Related would be rotate clockwise/counterclockwise as well as specific states, but then comes the point where we need gesture categories akin to @poire-z 's style tweaks.

**Comment by poire-z** [comment]
And already came the point we don't need that much :) and some devices don't support multi touch gestures - and multi swipes are already super enough and don't conflict with anything :)

**Comment by Biep** [comment]
I suppose ideally the command manager would offer the gestures available for the current device, and allow commands to be placed in menus and associated with any of those gestures.  In case the same gesture is associated with multiple commands, there should be an order of precedence.  E.g. swiping left-to-right jumps back; goes back a page; goes to the previous document - in that order of precedence.

Thank you, poire, for the tweaks!

**Comment by Frenzie** [comment]
I'm actually quite directly inspired by Opera, which easily allowed for exactly that. See <https://github.com/koreader/koreader/issues/1008#issuecomment-65050183>.

That screenshot only showed gesture chaining with a simple comma (`,`). Command chaining worked with an ampersand (`&`). Here's my custom keyboard setup:

![screenshot_2019-02-17_20-00-02](https://user-images.githubusercontent.com/202757/52917682-a6968b80-32ee-11e9-8343-e440d7990f4e.png)

Spiritual successor Vivaldi has a clever thing where you can record your own gesture, but you can't seem to chain commands. Otter, the other spiritual successor, does that much better.

**Comment by poire-z** [comment]
Another one that I would find useful: open the Skim document widget. Currently available via menu or with a long-press on a very tiny height over the bottom status bar (this tiny height is needed to not prevent highlighting a word in last line of book text).

(And that Gesture manager menu would probably be better placed in the `Screen>` menu than in the current `Device>` menu I think.)

**Comment by Frenzie** [comment]
Or screen under device?

**Comment by poire-z** [comment]
You mean `Device > Screen` ? Or just related vertical position in the Gear menu?
If the former, no :) Don't need to burry that useful `Screen>` too deep in a `Device > Screen >`

**Comment by Frenzie** [comment]
I meant the former, yes.

Wouldn't putting it under navigation make more sense?

**Comment by poire-z** [comment]
`Navigation`, currently, has stuff only related to key navigation (and is not displayed on my kobo).
`Screen` already has some stuff related to gesture (swipe for menu, disable double tap).
So, dunno :)
(But on kobo and the emulator, even if not the same items, that Gear menu has 10 items, and makes out one page - and I'm not fond of having more than one page in menus.)

**Comment by Frenzie** [comment]
>  I'm not fond of having more than one page in menus.

Me neither. That's exactly why I'd like to put more things in submenus. :-P

**Comment by poire-z** [comment]
For completeness, there could be some prev/next chapter actions. They are bound to onDoubleTapForward/Backward in readerrolling, but not in readerpaging, so it might need a bit of work.

**Comment by Biep** [comment]
I'd like to be able to bind some favourite books to gestures.  Whatever I have been doing, swiping ^ (think arrow to heaven) would bring up my Bible; v would bring up the childrens' bed-time book, and so on.

**Comment by Frenzie** [comment]
@poire-z Would you like to look into it? I know nothing about those. (They're also not in the skim dialog.)

**Comment by poire-z** [comment]
I may when I find the time.
But they are in skimtowidget.lua: button_chapter_next and button_chapter_prev , but they use some toc cooking that the SkimToWidget has already build to display the chapter marker on the progress bar... so not usable as is.

**Comment by Frenzie** [comment]
@poire-z `DoubleTapForward` seems like a much too specific name. I'd change it to something more generic like `GoToNextChapter` and then we can add the same event handler in `ReaderPaging`.

**Comment by poire-z** [comment]
Indeed :) ( I mention it because I saw you changed page forward to use `self.ui:handleEvent(Event:new("TapForward"))` - so you could have taken that shortcut for that too.)
So ok with a more generic rewrite.

**Comment by Frenzie** [comment]
That was mainly a quick workaround because that issue was *really, really* annoying. :blush: You're right that it should be changed.

**Comment by poire-z** [comment]
Well, about this next/prev chapter:
It looks as easy as what you are doing for tapForward: rename `ReaderRolling:onDoubleTapBackward` to `ReaderRolling:onGotoNextChapter`, use that in the registerTouchZones:
```diff
        {
            id = "double_tap_forward",
            ges = "double_tap",
            screen_zone = forward_double_tap_zone,
-           handler = function() return self:onDoubleTapForward() end
+           handler = function() return self:onGotoNextChapter() end
        },
```
(keep the `id = "double_tap_forward"`,, use the real function in handler=).

There's just the thing that there is nothing about chapters in readerpaging, not a single use of `self.ui.toc.` - so I guess that with PDF, people hardly need that, or TOC are often empty/untrustable.
So, if that's fine, I'll do that only for readerrolling - and leave the addition of next/prevChapter in readerpaging to those really interested by that :)

**Comment by Frenzie** [comment]
I'd say PDF TOC is generally fine when it exists. It's just that facsimile PDF/DjVu (e.g., Archive.org, Gallica) tends to have none.

tl;dr I might give it a look.

**Comment by poire-z** [comment]
I just have it ready, PR'ing.

**Comment by poire-z** [comment]
#4692 . Feel free to hack / re-order / add / squash / merge it.
